### ISSUE TITLE
update cycle is not topologically ordered

### ISSUE THREAD

**Issue by omar-azmi** [omar-azmi/tsignal_ts#4]
### problem:

I just realized that the `Context`'s [update cycle](https://github.com/omar-azmi/tsignal_ts/blob/3eff039d0bd644d898d331cece4feff58b2e4df1/src/context.ts#L139) mechanism is not actually topologically ordered.
the 3eff039d0bd644d898d331cece4feff58b2e4df1 commit adds the test file [`/test/simple_signal2.test.ts`](../blob/master/test/simple_signal2.test.ts), in which the traversal of ids is:
- expected traversal: A, B, D, G, C, E, H, F, I, J
- actual traversal by the update cycle: A, B, D, G, C, E, H,I, J, F

as you can see, signal `F` was supposed to be executed before `I` and `J`, but instead, it is pushed to the very end.

the directed acyclic graph looks as the following:
> A->[B,C] ; B->[D, F] ; C->[E, F] ; D->[F, G] ; E->[F, H] ; F->[I] ; G->[I] ; H->[I] ; I->[J] ;

```mermaid
---
title: "graph"
---
flowchart LR
	332796(("A")) --> 590207(("B"))
	332796 --> 763885(("C"))
	590207 --> 562158(("D"))
	763885 --> 932932(("E"))
	562158 --> 831043(("G"))
	831043 --> 450102(("I"))
	932932 --> 940461(("H"))
	940461 --> 450102
	590207 --> 270457(("F"))
	562158 --> 270457
	763885 --> 270457
	932932 --> 270457
	450102 --> 770803(("J"))
```

### reason:

`F` gets skipped in the middle because it is already [*visited*](https://github.com/omar-azmi/tsignal_ts/blob/3eff039d0bd644d898d331cece4feff58b2e4df1/src/context.ts#L173) once from either `B` or `D`, but its resolution is delayed because of its [*unvisited*](https://github.com/omar-azmi/tsignal_ts/blob/3eff039d0bd644d898d331cece4feff58b2e4df1/src/context.ts#L151) dependencies `C` and `E`.
however, when `C` and `E` are visited, they propagate [*forward*](https://github.com/omar-azmi/tsignal_ts/blob/3eff039d0bd644d898d331cece4feff58b2e4df1/src/context.ts#L173) to `H` and then `I`.
in `I`'s perspective, `F` has been resolved, since it has been [*visited*](https://github.com/omar-azmi/tsignal_ts/blob/3eff039d0bd644d898d331cece4feff58b2e4df1/src/context.ts#L140C8-L140C34). however, the reality is that while it has been visited, it was awaiting dependency resolution of `C` and `E`, and it hasn't been resolved yet.
the forward propagation model is indeed problematic. you should either stick backward resolution or forward propagation, but not both at the same time.

```mermaid
---
title: "the current incorrect traversal mechanism"
---
flowchart LR
	332796(("A")) --> 590207(("B"))
	590207 --> 562158(("D"))
	932932(("E")) --> 940461(("H"))
	940461 --> 450102(("I"))
	562158 --> 270457(("F"))
	270457 -.->|"backward"| 763885(("C"))
	763885 --> 932932
	450102 -.->|"backward"| 831043(("G"))
	450102 --> 770803(("J"))
	270457 -.-o|"get"| 450102
	562158 -.-o|"get"| 831043
```

**Comment by omar-azmi** [comment]
issue solved in the commit 82ce80d04f5ef0c7ca9af21048512f235eb7172c .

- [ ] I'd still like to add a good documentation in the readme on how the algorithm works.
- [ ] I also now need to inspect if the [`rmap`](https://github.com/omar-azmi/tsignal_ts/blob/6ef4118846d5ae76cfe1f8dc585225d91c024ff8/src/context.ts#L47) is even necessary anymore, since its main consumer was the old fake-topological update cycle, and the new one does not make use of it.

once the two tasks above are done, I'll close the issue.

**Comment by omar-azmi** [comment]
closing the issue, so that it shows up as resolved in the `v0.2.1` release.
